### DISCUSSION TITLE
How to override pytest logging behaviour

### DISCUSSION THREAD

**Discussion by mohit2152sharma** [pytest-dev/pytest#12035] [Q&A]
I have created a custom logging class and I am trying to write tests for it. The class has it's own formatter, for example something like ` "timestamp:%(asctime)s - logger_name:%(name)s - levelname:%(levelname)s - message:%(message)s"`. 

When i run the pytests, somehow the logs generated by this custom logger are captured in `caplog` with `basicConfig` format and on the `stdout` I can see the logs in the format set by custom logging class. 

Following are the code snippets of how I am doing it: 

```python
# this _core.py
class CustomLogger:
    def __init__(
        self,
        name: str = "root",
        level: int = logging.INFO,
        config: Optional[dict] = None,
    ):
        self.name = name
        self.level = level

        self.config = config or CustomDefaultConf()._create_default_conf_dict()
        logging.config.dictConfig(self.config)
        logging.setLoggerClass(CustomLoggerClass)
        self.logger = logging.getLogger(name)

# in the __init__.py I expose logger
logger = CustomLogger().logger

# in the tests file
from custom_logger import logger 

def test_dummy_func(): 
    logger.info("hello world")

# in the terminal I get the following output
---------------------------------------------------------------------------------------------- Captured stdout call ----------------------------------------------------------------------------------------------
timestamp:2024-02-26 19:56:38 - logger_name:custom_logger - levelname:INFO - time_taken:1 - message:Function dummy_func took 0:00:00.000163
----------------------------------------------------------------------------------------------- Captured log call ------------------------------------------------------------------------------------------------
INFO     custom_logger.py:42 Function dummy_func took 0:00:00.000163
```

How can I redirect the output of my custom logger to `"Captured log call"`?

**Comment by portyu9** [comment]
Hi @mohit2152sharma 

Those are two different handlers/output paths.

Your custom `StreamHandler` formats the record and writes that text to stdout, which is why you see your format under **Captured stdout**.

pytest's logging plugin captures the underlying `LogRecord` separately and formats it using **pytest's own formatter**, which is why **Captured log call** shows:

```text
INFO custom_logger.py:42 ...
```

If you want pytest's captured-log section to use your format, configure pytest's log formatter:

```ini
[pytest]
log_level = INFO
log_format = timestamp:%(asctime)s - logger_name:%(name)s - levelname:%(levelname)s - message:%(message)s
```

You can also specify it on the command line with `--log-format`.

If you do not want the duplicate **Captured stdout** section, remove/disable the application's stdout `StreamHandler` for the test configuration and let the logger propagate to pytest's capture handler.

One other important detail: `logging.config.dictConfig()` can replace the root logger's handlers. If you configure the root logger during a test, preserve pytest's existing handlers; otherwise `caplog` may stop capturing altogether.

So you don't redirect the already-formatted stdout text into `Captured log call`; configure pytest's logging formatter to format the same `LogRecord` the way you want.

Please mark this answer as accepted if it helped, thank you.
